import robokop_genetics.node_types as node_types
from robokop_genetics.simple_graph_components import SimpleNode
from robokop_genetics.util import Text, LoggingUtil
from math import ceil
from typing import Set
import logging
import requests


class ClinGenService(object):

    def __init__(self, log_file_path=None):
        self.logger = LoggingUtil.init_logging(__name__,
                                               logging.INFO,
                                               logFilePath=log_file_path)
        self.url = 'https://reg.genome.network/'
        self.synon_fields_param = 'fields=none+@id+externalRecords.dbSNP+externalRecords.ClinVarVariations+externalRecords.MyVariantInfo_hg38+genomicAlleles-genomicAlleles.referenceSequence'

    def get_batch_of_synonyms(self, variant_list, variant_format="hgvs"):
        # possible variant_format values not implemented yet
        # would only need to switch prefix for building synonym dictionary
        # id (CA ID or PA ID)
        # MyVariantInfo_hg19.id
        # MyVariantInfo_hg38.id
        # ExAC.id
        # gnomAD.id
        if not variant_list:
            return []

        separator = '\n'
        results_list = []
        batches = ceil(len(variant_list) / 2000)
        for i in range(batches):
            variant_subset = variant_list[i * 2000:i * 2000 + 2000]
            hgvs_pseudo_file = separator.join(variant_subset)
            query_url = f'{self.url}alleles?file={variant_format}&{self.synon_fields_param}'
            all_alleles_json = self.query_service(query_url, data=hgvs_pseudo_file)
            for index, allele_json in enumerate(all_alleles_json):
                results_list.append(self.parse_allele_json_for_synonyms(allele_json))
        return results_list

    def get_synonyms_by_caid(self, caid):
        synonyms = set()
        query_url = f'{self.url}allele/{caid}?{self.synon_fields_param}'
        allele_json = self.query_service(query_url)
        if allele_json:
            synonyms.update(self.parse_allele_json_for_synonyms(allele_json))
        return synonyms

    def get_synonyms_by_hgvs(self, hgvs_id):
        synonyms = set()
        query_url = f'{self.url}allele?hgvs={hgvs_id}&{self.synon_fields_param}'
        allele_json = self.query_service(query_url)
        if allele_json:
            synonyms.update(self.parse_allele_json_for_synonyms(allele_json))
        return synonyms

    def get_synonyms_by_rsid_with_sequence(self, rsid):
        if rsid.startswith('rs'):
            rsid = rsid[2:]

        return self.get_synonyms_by_parameter_matching('dbSNP.rs', rsid)

    def get_synonyms_by_other_ids(self, synonym_set: Set):
        # Just looking for 1 hit because any hit should return all of the available synonyms and caid if they exist
        hgvs_ids = Text.get_curies_by_prefix('HGVS', synonym_set)
        if hgvs_ids:
            for hgvs_id in hgvs_ids:
                synonyms = self.get_synonyms_by_hgvs(Text.un_curie(hgvs_id))
                if synonyms: return synonyms

        dbsnp_ids = Text.get_curies_by_prefix('DBSNP', synonym_set)
        if dbsnp_ids:
            for dbsnp_id in dbsnp_ids:
                rsid = Text.un_curie(dbsnp_id)
                if rsid.startswith('rs'):
                    synonyms = self.get_synonyms_by_parameter_matching('dbSNP.rs', rsid[2:])
                    if synonyms: return synonyms

        clinvar_ids = Text.get_curies_by_prefix('CLINVARVARIANT', synonym_set)
        if clinvar_ids:
            for clinvar_id in clinvar_ids:
                synonyms = self.get_synonyms_by_parameter_matching('ClinVar.variationId', Text.un_curie(clinvar_id))
                if synonyms: return synonyms

        myvariant_hg38_ids = Text.get_curies_by_prefix('MYVARIANT_HG38', synonym_set)
        if myvariant_hg38_ids:
            for myvariant_id in myvariant_hg38_ids:
                synonyms = self.get_synonyms_by_parameter_matching('MyVariantInfo_hg38.id', Text.un_curie(myvariant_id))
                if synonyms: return synonyms

        myvariant_hg19_ids = Text.get_curies_by_prefix('MYVARIANT_HG19', synonym_set)
        if myvariant_hg19_ids:
            for myvariant_id in myvariant_hg19_ids:
                synonyms = self.get_synonyms_by_parameter_matching('MyVariantInfo_hg19.id', Text.un_curie(myvariant_id))
                if synonyms: return synonyms

        return set()

    def get_synonyms_by_parameter_matching(self, url_param, url_param_value):
        synonyms = set()
        query_url = f'{self.url}alleles?{url_param}={url_param_value}&{self.synon_fields_param}'
        query_json = self.query_service(query_url)
        for allele_json in query_json:
            synonyms.update(self.parse_allele_json_for_synonyms(allele_json))
        return synonyms

    def parse_allele_json_for_synonyms(self, allele_json):
        synonyms = set()
        try:
            variant_caid = allele_json['@id'].rsplit('/', 1)[1]
        except (KeyError, IndexError):
            return synonyms

        synonyms.add(f'CAID:{variant_caid}')

        if 'genomicAlleles' in allele_json:
            try:
                for genomic_allele in allele_json['genomicAlleles']:
                    for hgvs_id in genomic_allele['hgvs']:
                        synonyms.add(f'HGVS:{hgvs_id}')
                    if 'referenceGenome' in genomic_allele and genomic_allele['referenceGenome'] == 'GRCh38':
                        if 'chromosome' in genomic_allele:
                            sequence = genomic_allele['coordinates'][0]['allele']
                            chromosome = genomic_allele['chromosome']
                            start_position = genomic_allele['coordinates'][0]['start']
                            end_position = genomic_allele['coordinates'][0]['end']
                            robokop_variant_id = f'HG38|{chromosome}|{start_position}|{end_position}|{sequence}'
                            synonyms.add(f'ROBO_VARIANT:{robokop_variant_id}')

            except KeyError as e:
                error_message = f'parsing sequence variant synonym - genomicAlleles KeyError for {variant_caid}: {e}'
                self.logger.info(error_message)

        if 'externalRecords' in allele_json:
            if 'MyVariantInfo_hg19' in allele_json['externalRecords']:
                for myvar_json in allele_json['externalRecords']['MyVariantInfo_hg19']:
                    myvariant_id = myvar_json['id']
                    synonyms.add(f'MYVARIANT_HG19:{myvariant_id}')

            if 'MyVariantInfo_hg38' in allele_json['externalRecords']:
                for myvar_json in allele_json['externalRecords']['MyVariantInfo_hg38']:
                    myvariant_id = myvar_json['id']
                    synonyms.add(f'MYVARIANT_HG38:{myvariant_id}')

            if 'dbSNP' in allele_json['externalRecords']:
                for dbsnp_json in allele_json['externalRecords']['dbSNP']:
                    variant_rsid = dbsnp_json['rs']
                    synonyms.add(f'DBSNP:rs{variant_rsid}')

            if 'ClinVarVariations' in allele_json['externalRecords']:
                for clinvar_json in allele_json['externalRecords']['ClinVarVariations']:
                    clinvar_id = clinvar_json['variationId']
                    synonyms.add(f'CLINVARVARIANT:{clinvar_id}')

        return synonyms

    def get_variants_by_region(self, reference_sequence_label, center_position, region_size):
        flanking_size = int(region_size / 2)
        begin = center_position - flanking_size
        if begin < 0: begin = 0
        end = center_position + flanking_size
        query_url_main = f'{self.url}alleles?refseq={reference_sequence_label}&begin={begin}&end={end}&fields=none+@id&limit=2000&skip='
        counter = 0
        return_results = []
        while True:
            query_url = f'{query_url_main}{counter}'
            query_results = self.query_service(query_url)
            if query_results:
                for allele_json in query_results:
                    if '@id' in allele_json:
                        id_split = allele_json['@id'].rsplit('/', 1)
                        if (len(id_split) > 1) and ('CA' in id_split[1]):
                            variant_caid = id_split[1]
                            variant_node = SimpleNode(id=f'CAID:{variant_caid}', type=node_types.SEQUENCE_VARIANT)
                            return_results.append(variant_node)
                counter += 2000
            else:
                break
        return return_results

    def query_service(self, query_url, data=None):
        if data:
            query_response = requests.post(query_url, data=data)
        else:
            query_response = requests.get(query_url)
        if query_response.status_code != 200:
            error_message = f'ClinGen returned a non-200 response({query_response.status_code}) calling ({query_url})'
            self.logger.warning(error_message)
            return {}
        else:
            query_json = query_response.json()
            return query_json
